File: services/metrics/domain/src/kui/metrics/domain/Throughput.scala
package kui.metrics.domain

import java.time.Instant

import scala.concurrent.duration.{DurationInt, FiniteDuration}

/** How far back a throughput question reaches, and at what resolution.
  *
  * Three ranges and not an arbitrary `from`/`to` pair, because the resolution has to be chosen with the
  * window: a thirty-day chart drawn at the five-minute step of a one-day chart is 8,640 points behind 900
  * pixels, and the browser would be downsampling a series KUI already had to hold in memory. Pairing each
  * window with its own step here is what keeps the bucket count roughly constant across all three.
  *
  * The `wire` string is the query-string spelling and is written out per case rather than derived from the
  * case name, for the reason `Resource.wire` is: renaming a case must never silently change what `?range=7d`
  * means to a browser that has already shipped.
  */
enum ThroughputRange(val wire: String, val window: FiniteDuration, val step: FiniteDuration) {

  case Last24Hours extends ThroughputRange("24h", 24.hours, 5.minutes)
  case Last7Days extends ThroughputRange("7d", 7.days, 1.hour)
  case Last30Days extends ThroughputRange("30d", 30.days, 6.hours)

  /** How many buckets a full series of this range holds. Constant per range, whatever was sampled: a range
    * that returned fewer buckets when the cluster was quiet would draw a shorter axis for a quiet hour.
    */
  def bucketCount: Int = (window.toSeconds / step.toSeconds).toInt
}

object ThroughputRange {

  val All: List[ThroughputRange] = List(Last24Hours, Last7Days, Last30Days)

  /** The default a caller who names no range gets. Stated once, here, so the contract's default and any later
    * caller's fallback cannot be two different windows labelled with one word.
    */
  val Default: ThroughputRange = Last24Hours

  /** Parses the query-string spelling. `None` for anything else, deliberately: falling back to the default
    * would answer a request for `?range=90d` with 24 hours of data under a label the caller chose, which is
    * the one failure a chart cannot show its reader.
    */
  def fromWire(raw: String): Option[ThroughputRange] = {
    val normalised = raw.trim.toLowerCase
    All.find(_.wire == normalised)
  }

  given CanEqual[ThroughputRange, ThroughputRange] = CanEqual.derived
}

/** One reading of a cluster's throughput, at the moment the source was scraped.
  *
  * Rates rather than counters. A broker publishes monotonic totals and turning two totals into a rate needs
  * both samples and the gap between them, which is the adapter's arithmetic (ADR-050); by the time a value is
  * a `ThroughputSample` that arithmetic has happened, and a consumer of this type can never accidentally
  * chart a counter that resets when a broker restarts.
  *
  * ==Each rate is separately optional, because an exporter publishes each of them separately==
  *
  * A JMX exporter is configured with a whitelist, and a deployment that publishes `BytesInPerSec` and
  * `BytesOutPerSec` without `MessagesInPerSec` is an ordinary configuration rather than a broken one. Making
  * the three rates one all-or-nothing tuple would mean refusing the whole card — the two rates the Traffic
  * screen actually draws included — because a third one nobody asked for was absent. `None` here means the
  * same thing it means one type down: *not measured*, never zero.
  */
final case class ThroughputSample(
    at: Instant,
    bytesInPerSecond: Option[Double],
    bytesOutPerSecond: Option[Double],
    recordsPerSecond: Option[Double]
) {

  /** True when the scrape produced no rate at all. A source answering these is a source answering nothing,
    * and the adapter refuses rather than filing an empty sample that would read as a measured gap.
    */
  def isEmpty: Boolean =
    bytesInPerSecond.isEmpty && bytesOutPerSecond.isEmpty && recordsPerSecond.isEmpty
}

object ThroughputSample {
  given CanEqual[ThroughputSample, ThroughputSample] = CanEqual.derived
}

/** One step of the time axis, and what was measured over it — or nothing, when nothing was.
  *
  * Every rate is an `Option` and every `None` means **not measured**. That is the single rule this whole type
  * exists to carry: a bucket KUI did not sample is a gap in the chart, and rendering it as `0` is a claim
  * that the cluster was idle. An operator looking at an idle-looking hour goes and asks why their producers
  * stopped; the honest answer was that KUI was restarting.
  */
final case class ThroughputBucket(
    startingAt: Instant,
    bytesInPerSecond: Option[Double],
    bytesOutPerSecond: Option[Double],
    recordsPerSecond: Option[Double]
) {

  /** True when nothing at all was measured over this step. A bucket of measured zeroes is not absent. */
  def isAbsent: Boolean =
    bytesInPerSecond.isEmpty && bytesOutPerSecond.isEmpty && recordsPerSecond.isEmpty
}

object ThroughputBucket {

  /** The bucket for a step nothing was sampled in. */
  def absentAt(startingAt: Instant): ThroughputBucket = ThroughputBucket(startingAt, None, None, None)

  given CanEqual[ThroughputBucket, ThroughputBucket] = CanEqual.derived
}

/** A cluster's throughput over one range, bucket by bucket, ending at a stated instant.
  *
  * `from` and `to` are carried rather than left for the reader to derive from the buckets, because a series
  * whose every bucket is absent still has to draw an axis: without them a chart with no data would have no
  * width and would render as an empty box rather than as an empty *day*.
  */
final case class ThroughputSeries(
    range: ThroughputRange,
    from: Instant,
    to: Instant,
    buckets: List[ThroughputBucket]
)

object ThroughputSeries {

  given CanEqual[ThroughputSeries, ThroughputSeries] = CanEqual.derived

  /** Folds raw samples into the range's buckets, ending at `endingAt`.
    *
    * This is the domain rule of the whole service, and it is stated here rather than in whichever adapter M7
    * writes so that a JMX source and a Prometheus source cannot disagree about what a gap is.
    *
    * Three things it guarantees, each of which has a wrong answer that reaches a screen:
    *
    *   - a bucket no sample fell into is **absent**, never zero;
    *   - a bucket whose samples were all zero is `Some(0.0)`, which is a measured quiet cluster and a
    *     different fact from the one above;
    *   - the boundaries are floored to the step against the epoch, so two requests a minute apart return the
    *     same buckets. Anchoring them to "now" instead would shift every boundary between two polls and make
    *     a live chart shiver even while the cluster's traffic was constant.
    *
    * Samples inside one bucket are averaged. A bucket is a step of wall-clock time and the value drawn for it
    * is the rate that held over it; taking the last sample instead would let one spike at 12:04 stand for the
    * whole of 12:00–12:05.
    */
  def over(range: ThroughputRange, endingAt: Instant, samples: List[ThroughputSample]): ThroughputSeries = {
    val stepSeconds = range.step.toSeconds
    val to = floorTo(endingAt, stepSeconds).plusSeconds(stepSeconds)
    val from = to.minusSeconds(range.window.toSeconds)

    val byBucket: Map[Instant, List[ThroughputSample]] =
      samples
        .filter(sample => !sample.at.isBefore(from) && sample.at.isBefore(to))
        .groupBy(sample => floorTo(sample.at, stepSeconds))

    val buckets = boundaries(from, to, stepSeconds).map { start =>
      // Each rate is folded over the samples that carried *it*, not over the samples in the bucket. An
      // exporter that started publishing `MessagesInPerSec` halfway through an hour must not make the
      // bytes it published all hour read as absent, and must not make the first half read as measured.
      val inBucket = byBucket.getOrElse(start, Nil)

      ThroughputBucket(
        startingAt = start,
        bytesInPerSecond = mean(inBucket.flatMap(_.bytesInPerSecond)),
        bytesOutPerSecond = mean(inBucket.flatMap(_.bytesOutPerSecond)),
        recordsPerSecond = mean(inBucket.flatMap(_.recordsPerSecond))
      )
    }

    ThroughputSeries(range, from, to, buckets)
  }

  /** The full axis with nothing on it: every bucket present, every rate absent.
    *
    * The rendering for a cluster KUI can reach but has never managed to sample. It is a *measured* answer —
    * "we looked and have nothing for this window" — and it is deliberately not the same value as no series at
    * all, which is what a cluster with no source configured produces.
    */
  def absent(range: ThroughputRange, endingAt: Instant): ThroughputSeries =
    over(range, endingAt, Nil)

  private def boundaries(from: Instant, to: Instant, stepSeconds: Long): List[Instant] =
    Iterator
      .iterate(from)(_.plusSeconds(stepSeconds))
      .takeWhile(_.isBefore(to))
      .toList

  /** The start of the step `at` falls in, measured from the epoch rather than from the request. */
  private def floorTo(at: Instant, stepSeconds: Long): Instant =
    Instant.ofEpochSecond(Math.floorDiv(at.getEpochSecond, stepSeconds) * stepSeconds)

  /** `None` for a rate nothing in the bucket measured — which is the one arithmetic mistake this file exists
    * to prevent, since `0.0 / 0` is `NaN` and a `NaN` serialises to `null` by a route nobody chose.
    */
  private def mean(values: List[Double]): Option[Double] =
    Option.when(values.nonEmpty)(values.sum / values.size)
}

File: frontend/packages/shell/src/overview/throughput.ts
/**
 * The Throughput card's data: the wire, the range, and the fold that turns buckets into a chart.
 *
 * ## The wire this file codes against
 *
 * `GET /api/v1/clusters/{clusterId}/metrics/throughput?range=24h|7d|30d` answers a `Section`-wrapped
 * `{throughput: {status, data}}`. When the status carries data, `data` is
 * `{range, from, to, stepSeconds, buckets}` and each bucket is
 * `{startingAt, bytesInPerSecond, bytesOutPerSecond, recordsPerSecond}` with every rate nullable.
 *
 * **A null rate means the step was never sampled.** It is not a zero, and the whole of this module
 * exists to keep those two apart from the moment the JSON is decoded to the moment a bar is drawn.
 * A zero is a measured fact about a quiet cluster; a null is KUI admitting it was not looking. An
 * operator shown a flat line for an hour KUI spent restarting goes and asks their producers why
 * they stopped, which is the most expensive wrong answer this screen can give.
 *
 * ## Why the bucket key is read under two names
 *
 * The committed contract (`services/metrics/contract/.../ThroughputDtos.scala`) calls it
 * `startingAt`; the wave's own contract table writes the same field `at`. They are the same field
 * and only one of them can be on the wire, so both are read and neither is preferred in code that
 * has to keep working across the change. This is not defensive decoding in general — the rest of
 * the payload is read exactly as the Scala writes it — it is one name that is written down twice.
 *
 * ## Why the buckets are drawn one for one and never re-bucketed here
 *
 * `SCREENS-V4.md` §4.1 draws twenty-four paired columns and a 24h range answers 288 of them, so the
 * obvious move is to average twelve buckets into each drawn column. It is the wrong move: a column
 * folded from twelve steps of which one was sampled draws as *measured*, and the eleven gaps inside
 * it disappear. Re-bucketing in the browser would hide exactly the thing this card is for. So every
 * bucket the server sent gets a column, `BarChart` narrows the bars to fit, and the gaps stay where
 * they happened.
 */

import type { KuiApiClient } from "@kui/api";
import { decodeSection } from "@kui/api";
import { apiFailure, fromSection, type Fetched } from "@kui/kernel";

/** The three windows the endpoint offers, in the order the selector draws them. */
export const THROUGHPUT_RANGES = ["24h", "7d", "30d"] as const;

export type ThroughputRange = (typeof THROUGHPUT_RANGES)[number];

/**
 * The range a request with no `?range=` gets, and the one the card opens on.
 *
 * The same value `ThroughputRange.Default` holds on the server. If the two ever disagree the card
 * asks for one window and labels it with another, so the agreement is asserted rather than assumed:
 * see `throughput.test.ts`.
 */
export const DEFAULT_THROUGHPUT_RANGE: ThroughputRange = "24h";

/** The search parameter the selector writes. In the address so a colleague can be sent one. */
export const RANGE_PARAM = "range";

/**
 * How many buckets a full series of each range holds — 24h/5min, 7d/1h, 30d/6h.
 *
 * A mirror of `ThroughputRange.bucketCount`, and it is used **only to check the answer**, never to
 * build the axis. The axis is always the buckets the server actually sent; this table is what lets
 * the card notice that it was sent fewer and say so, which is the browser's half of the rule that a
 * quiet window must draw the same axis as a busy one.
 */
export const BUCKET_COUNT: Readonly<Record<ThroughputRange, number>> = {
  "24h": 288,
  "7d": 168,
  "30d": 120,
};

/** What the segment or query parameter means. An unknown spelling is the default, not an error. */
export function throughputRange(raw: string | undefined): ThroughputRange {
  return THROUGHPUT_RANGES.find((range) => range === raw) ?? DEFAULT_THROUGHPUT_RANGE;
}

/* --- The wire, as far as this card reads it ---------------------------------------------------- */

/**
 * One step of the axis.
 *
 * Every rate is `number | null | undefined` and all three of those mean the same thing here: not
 * measured. `null` is what the server writes, `undefined` is what an older build that omitted the
 * field would leave, and neither may become a zero.
 */
export interface ThroughputBucket {
  readonly startingAt?: string | undefined;
  /** The other spelling of `startingAt`. See the header. */
  readonly at?: string | undefined;
  readonly bytesInPerSecond?: number | null | undefined;
  readonly bytesOutPerSecond?: number | null | undefined;
}

export interface ThroughputSeries {
  readonly range?: string | undefined;
  readonly from?: string | undefined;
  readonly to?: string | undefined;
  /** The width of one bucket. Used for the sentence that counts the gaps, and for nothing else. */
  readonly stepSeconds?: number | undefined;
  readonly buckets?: readonly ThroughputBucket[] | undefined;
}

/* --- The chart ---------------------------------------------------------------------------------- */

/** A run of consecutive buckets that were all measured, or all not. Drawn as the coverage strip. */
export interface CoverageRun {
  readonly measured: boolean;
  readonly length: number;
}

/** Everything the card draws, decided here so the component is arrangement. */
export interface ThroughputChart {
  /** One label per bucket. Its length is the axis, and it is the server's count, never a target. */
  readonly categories: readonly string[];
  /** Bytes in per second, per bucket. `null` is a gap. */
  readonly produce: readonly (number | null)[];
  /** Bytes out per second, per bucket. `null` is a gap. */
  readonly consume: readonly (number | null)[];
  /** Which categories get a printed tick: five, as `SCREENS-V4.md` §4.1 draws them. */
  readonly ticks: readonly number[];
  /** The gaps, in drawing order, so a reader can see *where* KUI was not looking. */
  readonly coverage: readonly CoverageRun[];
  /** How many buckets carried neither rate. */
  readonly absentBuckets: number;
  /** The newest bucket that carried a rate, for the legend chips. `undefined` when none did. */
  readonly latest: { readonly produce: number | null; readonly consume: number | null } | undefined;
  /** The sentence under the chart, or `undefined` when there is nothing to say. */
  readonly caption: string | undefined;
}

/** A rate the server sent, or `null` when it sent nothing. Never `0` for an absent measurement. */
function rate(value: number | null | undefined): number | null {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

/**
 * The category label for one bucket, in the browser's own zone.
 *
 * The zone is the browser's because nothing in KUI stores a timezone preference — `SettingsPage`'s
 * header says why there is no control for one — so a label in any other zone would be a claim the
 * rest of the product does not make. Each range gets the coarsest label that still separates two
 * adjacent buckets: minutes over a day, the weekday over a week, the date over a month.
 */
function labelFor(instant: string | undefined, range: ThroughputRange): string {
  if (instant === undefined) return "";
  const at = new Date(instant);
  if (Number.isNaN(at.getTime())) return "";
  switch (range) {
    case "24h":
      return at.toLocaleTimeString(undefined, { hour: "2-digit", minute: "2-digit" });
    case "7d":
      return at.toLocaleString(undefined, { weekday: "short", hour: "2-digit", minute: "2-digit" });
    case "30d":
      return at.toLocaleDateString(undefined, { day: "numeric", month: "short" });
  }
}

/**
 * Five ticks rather than `defaultTicks`'s three.
 *
 * §4.1 labels the throughput axis `00:00 · 06:00 · 12:00 · 18:00 · now`, which is four quarters and
 * an end. The kernel's default is first, middle and last, which is the right default for a card
 * narrow enough that five labels would collide and the wrong one for the widest card on the page.
 */
export function axisTicks(count: number): readonly number[] {
  if (count <= 0) return [];
  if (count <= 5) return Array.from({ length: count }, (_unused, index) => index);
  const last = count - 1;
  return [0, Math.round(last * 0.25), Math.round(last * 0.5), Math.round(last * 0.75), last];
}

/**
 * The gaps, as runs.
 *
 * A run rather than a cell per bucket, because a 30-day series is 120 cells and a 24-hour one is
 * 288, and the strip has to align with a chart that divides its width equally between buckets — so
 * a run of `n` buckets is a box with `flex-grow: n` and the alignment is arithmetic rather than a
 * measurement. A bucket is measured when *either* rate was measured: a source that publishes bytes
 * in and not bytes out is an ordinary exporter configuration, and calling that step unmeasured
 * would report a gap that is not there.
 */
export function coverageRuns(
  produce: readonly (number | null)[],
  consume: readonly (number | null)[],
): readonly CoverageRun[] {
  const runs: CoverageRun[] = [];
  for (let index = 0; index < produce.length; index += 1) {
    const measured = produce[index] !== null || (consume[index] ?? null) !== null;
    const last = runs[runs.length - 1];
    if (last !== undefined && last.measured === measured) {
      runs[runs.length - 1] = { measured, length: last.length + 1 };
    } else {
      runs.push({ measured, length: 1 });
    }
  }
  return runs;
}

/** How long one step is, in words, from the width the server sent. `undefined` when it sent none. */
function stepWords(stepSeconds: number | undefined): string | undefined {
  if (stepSeconds === undefined || !Number.isFinite(stepSeconds) || stepSeconds <= 0) return undefined;
  if (stepSeconds % 3600 === 0) {
    const hours = stepSeconds / 3600;
    return hours === 1 ? "one-hour" : `${hours}-hour`;
  }
  if (stepSeconds % 60 === 0) {
    const minutes = stepSeconds / 60;
    return minutes === 1 ? "one-minute" : `${minutes}-minute`;
  }
  return `${stepSeconds}-second`;
}

/**
 * The sentence under the chart.
 *
 * Two facts can need saying and they are said in one line rather than two, because a card with two
 * captions reads as a card that has gone wrong. Neither is decoration: the gap count is what makes
 * a blank column legible as "not measured" rather than as "nothing happened", and the short-axis
 * clause is the browser noticing that the server broke the one rule this chart's shape rests on —
 * that a range's bucket count is constant whatever was sampled.
 */
function captionFor(
  range: ThroughputRange,
  drawn: number,
  absent: number,
  stepSeconds: number | undefined,
): string | undefined {
  const parts: string[] = [];
  const expected = BUCKET_COUNT[range];
  if (drawn !== expected) {
    parts.push(
      `This build answered ${drawn} steps where a ${range} range holds ${expected}, so the axis is ` +
        `shorter than the window it is labelled with.`,
    );
  }
  if (absent > 0) {
    const step = stepWords(stepSeconds);
    const steps = step === undefined ? "steps" : `${step} steps`;
    parts.push(
      `${absent} of the ${drawn} ${steps} in this window were never sampled, and are drawn blank ` +
        `rather than as a rate of zero.`,
    );
  }
  return parts.length === 0 ? undefined : parts.join(" ");
}

/**
 * The wire's series as the picture, and the one line in this file that matters.
 *
 * `rate()` is what keeps a gap a gap. Replacing either of its two calls with `?? 0` turns every
 * unsampled step into a measured quiet one, and the chart, the coverage strip and the hidden data
 * table all stop being able to tell the difference — which is the mutation this card is gated on.
 */
export function throughputChart(series: ThroughputSeries, range: ThroughputRange): ThroughputChart {
  const buckets = series.buckets ?? [];
  const categories = buckets.map((bucket) => labelFor(bucket.startingAt ?? bucket.at, range));
  const produce = buckets.map((bucket) => rate(bucket.bytesInPerSecond));
  const consume = buckets.map((bucket) => rate(bucket.bytesOutPerSecond));
  const coverage = coverageRuns(produce, consume);
  const absentBuckets = coverage.reduce((sum, run) => (run.measured ? sum : sum + run.length), 0);

  /* The newest bucket that measured anything, which is what the legend chips print. Newest rather
     than an average over the window: §3.1 puts the *current* value in the chip, and an average
     across a day is a number nobody can act on. */
  let latest: ThroughputChart["latest"] = undefined;
  for (let index = buckets.length - 1; index >= 0; index -= 1) {
    const producedAt = produce[index] ?? null;
    const consumedAt = consume[index] ?? null;
    if (producedAt !== null || consumedAt !== null) {
      latest = { produce: producedAt, consume: consumedAt };
      break;
    }
  }

  return {
    categories,
    produce,
    consume,
    ticks: axisTicks(categories.length),
    coverage,
    absentBuckets,
    latest,
    caption: captionFor(range, buckets.length, absentBuckets, series.stepSeconds),
  };
}

/**
 * Whether this series measured anything at all.
 *
 * Cheap on purpose: the voice line above the tab asks this on every render, and building the whole
 * chart to answer it would fold 288 buckets into three arrays to find out whether one of them held
 * a number. It stops at the first measured rate.
 */
export function hasMeasuredBucket(series: ThroughputSeries): boolean {
  return (series.buckets ?? []).some(
    (bucket) => rate(bucket.bytesInPerSecond) !== null || rate(bucket.bytesOutPerSecond) !== null,
  );
}

/* --- Asking for it ------------------------------------------------------------------------------ */

/**
 * The key one request is cached under.
 *
 * Both the cluster and the range are in it, which is `useQuery`'s whole contract: a range left out
 * would mean the selector changed the address, changed the label and drew the previous window's
 * bars out of the cache.
 */
export function throughputKey(clusterId: string, range: ThroughputRange): string {
  return `metrics-throughput:${clusterId}:${range}`;
}

/**
 * Asks for one cluster's throughput.
 *
 * Never rejects, by the same rule the rest of `@kui/api` follows. The four section statuses that do
 * not carry data become the four `Fetched` cases that say why, and `not_configured` in particular
 * stays its own case all the way to the card: it is the common, fully supported answer for a
 * deployment that has configured no exporter, and turning it into a failure would put a red panel
 * and a retry button in front of every operator who never asked for metrics.
 */
export async function fetchThroughput(
  api: KuiApiClient,
  clusterId: string,
  range: ThroughputRange,
): Promise<Fetched<ThroughputSeries>> {
  const answer = await api.get("/api/v1/clusters/{clusterId}/metrics/throughput", {
    params: { path: { clusterId }, query: { range } },
  });
  if (!answer.ok) return apiFailure(answer.error);

  const body = answer.value as { throughput?: unknown } | null;
  if (body === null || body.throughput === undefined) {
    return {
      kind: "failed",
      message: "KUI could not read the throughput: the server sent something other than a series.",
      code: "UNREADABLE_BODY",
    };
  }
  return fromSection(decodeSection<ThroughputSeries>(body.throughput), (data) => data);
}
